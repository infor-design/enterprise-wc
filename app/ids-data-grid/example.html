<ids-layout-grid>
  <ids-text font-size="12" type="h1">Data Grid</ids-text>
</ids-layout-grid>
<ids-layout-grid auto="true">
  <ids-layout-grid-cell>
    <ids-data-grid id="data-grid-1" virtual-scroll="false" label="Books" row-height="medium">
    </ids-data-grid>
  </ids-layout-grid-cell>
</ids-layout-grid>
