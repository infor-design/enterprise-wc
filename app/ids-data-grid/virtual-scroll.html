{{> ../layouts/head.html }}
<ids-layout-grid>
  <ids-text font-size="12" type="h1">Data Grid</ids-text>
</ids-layout-grid>
<ids-layout-grid auto="true">
  <ids-layout-grid-cell>
    <ids-data-grid id="data-grid-1" virtual-scroll="true" alternate-row-shading="true" row-height="small">
    </ids-data-grid>
  </ids-layout-grid-cell>
</ids-layout-grid>
{{> ../layouts/footer.html }}
