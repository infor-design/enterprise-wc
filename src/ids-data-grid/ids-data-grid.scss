@import '../ids-base/ids-base';

$cell-border-color: var(--ids-color-palette-slate-30);
$header-bg-color: var(--ids-color-palette-slate-80);

:host {
  --ids-data-grid-column-widths: repeat(1, minmax(110px, 1fr));
}

.ids-data-grid {
  // Outside border and Background and sizing
  @include bg-white();
  @include border-1();
  @include border-solid();
  @include rounded-default();
  @include w-full();

  border-collapse: collapse;
  border-color: $cell-border-color;
  overflow: auto;

  // Font Info
  @include antialiased();
  @include font-sans();
  @include text-16();

  // Row Styles
  .ids-data-grid-row {
    @include w-full();

    display: grid;
    grid-template-columns: var(--ids-data-grid-column-widths); // Filled in dynamically
  }

  // Row Height Feature
  .ids-data-grid-body .ids-data-grid-row {
    height: 50px;
  }

  &[data-row-height='medium'] .ids-data-grid-body .ids-data-grid-row {
    height: 40px;
  }

  &[data-row-height='small'] .ids-data-grid-body .ids-data-grid-row {
    height: 35px;
  }

  &[data-row-height='extra-small'] .ids-data-grid-body .ids-data-grid-row {
    height: 30px;
  }

  // Row Alternate Row Setting Feature
  &.alt-row-shading .ids-data-grid-body [role='row']:nth-child(odd) {
    @include bg-slate-10();
  }

  // Header Styles
  .ids-data-grid-header {
    @include z-2();
    @include sticky();

    top: 0;
  }

  // Header Cell Area
  .ids-data-grid-header-cell {
    @include border-0();
    @include border-l-1();
    @include border-solid();
    @include flex();

    background-color: $header-bg-color;
    border-color: $cell-border-color;
    min-height: 35px;

    // Hide first border
    &:first-child {
      border-left: 0;
    }
  }

  .ids-data-grid-header-text {
    @include block();
    @include px-16();
    @include py-8();
    @include font-bold();
    @include text-14();
    @include text-white();

    overflow: hidden;
    text-align: left;
    text-overflow: ellipsis;
    user-select: none;
    white-space: nowrap;
  }

  // Header sort styles
  .sort-indicator {
    @include relative();

    height: 16px;
    left: -6px;
    width: 18px;
  }

  // Sort Hover states
  .is-sortable {
    ids-icon {
      color: var(--ids-color-palette-slate-50);
      opacity: 0;
      transition: opacity 0.3s cubic-bezier(0.17, 0.04, 0.03, 0.94) 0s;
    }

    ids-icon:first-child {
      @include relative();

      top: 4px;
      transform: rotate(180deg);
    }

    ids-icon:last-child {
      @include relative();

      top: -9px;
    }

    &[aria-sort='descending'] ids-icon {
      &:first-child {
        opacity: 1;
      }

      &:last-child {
        color: var(--ids-color-palette-white);
        opacity: 1;
      }
    }

    // stylelint-disable
    &[aria-sort='ascending'] ids-icon {
      &:first-child {
        color: var(--ids-color-palette-white);
        opacity: 1;
      }

      &:last-child {
        opacity: 1;
      }
    }

    &:hover {
      cursor: pointer;

      .sort-indicator ids-icon {
        opacity: 1;
      }
    }
  }

  // Cell Styles
  .ids-data-grid-cell {
    // Border and Background
    @include border-0();
    @include border-l-1();
    @include border-t-1();
    @include border-solid();
    @include inline-flex();

    align-items: center;
    border-color: $cell-border-color;
    overflow: hidden;
    white-space: nowrap;

    // Hide first border
    &:first-child {
      border-left: 0;
    }

    // Text ellipsis
    .text-ellipsis {
      @include block();
      @include px-16();

      overflow: hidden;
      text-overflow: ellipsis;
    }

    // Focus State
    &:focus {
      @include shadow();
      outline: 1px solid var(--ids-color-palette-azure-60);
      outline-offset: -1px;
    }
  }

  // Standalone Css
  tr.ids-data-grid-header {
    position: initial;
  }

  tr.ids-data-grid-row {
    @include table-row();

    height: 50px;
  }

  td.ids-data-grid-cell,
  th.ids-data-grid-header-cell {
    @include table-cell();
  }
}
