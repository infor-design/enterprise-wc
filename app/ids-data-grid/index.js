import { IdsDataGrid } from '../../src/ids-data-grid/ids-data-grid';

// Example for populating the DataGrid
const dataGrid = document.querySelector('#data-grid-1');

// Do an ajax request
const xmlhttp = new XMLHttpRequest();
const url = '/api/books';
const columns = [];

// Set up columns
columns.push({
  id: 'selectionCheckbox',
  sortable: false,
  resizable: false,
  formatter: dataGrid.formatters.text,
  align: 'center',
  width: 20
});
columns.push({
  id: 'book',
  name: 'Book',
  field: 'book',
  formatter: dataGrid.formatters.text,
  sortable: true,
  width: 65
});
columns.push({
  id: 'description',
  name: 'Description',
  field: 'description',
  sortable: true,
  formatter: dataGrid.formatters.text
});
columns.push({
  id: 'ledger',
  name: 'Ledger',
  field: 'ledger',
  formatter: dataGrid.formatters.text
});
columns.push({
  id: 'bookCurrency',
  name: 'Book Currency',
  field: 'bookCurrency',
  formatter: dataGrid.formatters.text
});
columns.push({
  id: 'transactionCurrency',
  name: 'Transaction Currency',
  field: 'transactionCurrency',
  formatter: dataGrid.formatters.text,
});
columns.push({
  id: 'postHistory',
  name: 'Post History',
  field: 'postHistory',
  formatter: dataGrid.formatters.text
});
columns.push({
  id: 'active',
  name: 'Active',
  field: 'active',
  formatter: dataGrid.formatters.text
});
columns.push({
  id: 'convention',
  name: 'Convention',
  field: 'convention',
  formatter: dataGrid.formatters.text
});
columns.push({
  id: 'methodSwitch',
  name: 'Method Switch',
  field: 'methodSwitch',
  formatter: dataGrid.formatters.text,
  filterType: 'select'
});
columns.push({
  id: 'trackDeprecationHistory',
  name: 'Track Deprecation History',
  field: 'trackDeprecationHistory',
  formatter: dataGrid.formatters.dropdown
});
columns.push({
  id: 'useForEmployee',
  name: 'Use For Employee',
  field: 'useForEmployee',
  formatter: dataGrid.formatters.dropdown
});
columns.push({
  id: 'deprecationHistory',
  name: 'Deprecation History',
  field: 'deprecationHistory',
  formatter: dataGrid.formatters.dropdown
});

xmlhttp.onreadystatechange = function onreadystatechange() {
  if (this.readyState === 4 && this.status === 200) {
    dataGrid.data = JSON.parse(this.responseText);
    dataGrid.columns = columns;
  }
};

// Execute the request
xmlhttp.open('GET', url, true);
xmlhttp.send();
