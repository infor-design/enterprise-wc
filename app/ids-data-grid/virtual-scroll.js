import { IdsDataGrid } from '../../src/ids-data-grid/ids-data-grid';

// Example for populating the DataGrid
const dataGrid = document.querySelector('#data-grid-1');

// Do an ajax request
const xmlhttp = new XMLHttpRequest();
const url = '/api/products';
const columns = [];

// Set up columns
columns.push({
  id: 'id',
  name: 'ID',
  field: 'id',
  formatter: dataGrid.formatters.text,
  width: 80,
  sortable: true
});
columns.push({
  id: 'color',
  name: 'Color',
  field: 'color',
  formatter: dataGrid.formatters.text,
  sortable: true
});
columns.push({
  id: 'inStock',
  name: 'In Stock',
  field: 'inStock',
  formatter: dataGrid.formatters.text,
  sortable: true
});
columns.push({
  id: 'productId',
  name: 'Product Id',
  field: 'productId',
  formatter: dataGrid.formatters.text,
  sortable: true
});
columns.push({
  id: 'productName',
  name: 'Product Name',
  field: 'productName',
  formatter: dataGrid.formatters.text,
  sortable: true
});
columns.push({
  id: 'unitPrice',
  name: 'Unit Price',
  field: 'unitPrice',
  formatter: dataGrid.formatters.text,
  sortable: true
});
columns.push({
  id: 'units',
  name: 'Units',
  field: 'units',
  formatter: dataGrid.formatters.text,
  sortable: true
});

xmlhttp.onreadystatechange = function onreadystatechange() {
  if (this.readyState === 4 && this.status === 200) {
    dataGrid.columns = columns;
    dataGrid.data = JSON.parse(this.responseText);
  }
};

// Execute the request
xmlhttp.open('GET', url, true);
xmlhttp.send();
