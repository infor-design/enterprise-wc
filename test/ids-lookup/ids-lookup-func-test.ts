/**
 * @jest-environment jsdom
 */

import '../helpers/resize-observer-mock';
import '../helpers/match-media-mock';
import '../helpers/canvas-mock';
import createFromTemplate from '../helpers/create-from-template';
import waitForTimeout from '../helpers/wait-for-timeout';
import dataset from '../../src/assets/data/books.json';

import '../../src/components/ids-data-grid/ids-data-grid';
import IdsDataGridFormatters from '../../src/components/ids-data-grid/ids-data-grid-formatters';
import IdsLookup from '../../src/components/ids-lookup/ids-lookup';

describe('IdsLookup Component', () => {
  let lookup: any;
  const formatters = new IdsDataGridFormatters();

  const columns = () => {
    const cols = [];
    // Set up columns
    cols.push({
      id: 'selectionCheckbox',
      sortable: false,
      resizable: false,
      formatter: formatters.selectionCheckbox,
      align: 'center'
    });
    cols.push({
      id: 'rowNumber',
      name: '#',
      formatter: formatters.rowNumber,
      sortable: false,
      readonly: true,
      width: 65
    });
    cols.push({
      id: 'description',
      name: 'Description',
      field: 'description',
      sortable: true,
      formatter: formatters.text
    });
    cols.push({
      id: 'ledger',
      name: 'Ledger',
      field: 'ledger',
      formatter: formatters.text
    });
    cols.push({
      id: 'price',
      name: 'Price',
      field: 'price',
      formatter: formatters.decimal,
      formatOptions: { locale: 'en-US' } // Data Values are in en-US
    });
    cols.push({
      id: 'bookCurrency',
      name: 'Currency',
      field: 'bookCurrency',
      formatter: formatters.text
    });
    return cols;
  };

  const createMultiSelectLookup = async () => {
    lookup = await createFromTemplate(lookup, `<ids-lookup id="lookup-1" label="Normal Lookup" title="Select an Item" field="description"></ids-lookup>`);
    lookup.dataGridSettings = {
      rowSelection: 'multiple'
    };
    lookup.columns = columns();
    lookup.data = dataset;
    return lookup;
  };

  beforeEach(async () => {
    lookup = await createFromTemplate(lookup, `<ids-lookup id="lookup-1" label="Normal Lookup"></ids-lookup>`);
  });

  afterEach(async () => {
    document.body.innerHTML = '';
  });

  it('renders empty dropdown with no errors', () => {
    const errors = jest.spyOn(global.console, 'error');
    lookup.remove();
    const elem: any = new IdsLookup();
    document.body.appendChild(elem);
    elem.remove();
    expect(document.querySelectorAll('ids-lookup').length).toEqual(0);
    expect(errors).not.toHaveBeenCalled();
  });

  it('renders correctly', () => {
    expect(lookup.outerHTML).toMatchSnapshot();
    expect(lookup.shadowRoot.innerHTML).toMatchSnapshot();
  });

  it('renders with disabled', () => {
    lookup = createFromTemplate(lookup, `<ids-lookup id="lookup-1" disabled="true" label="Test"></ids-lookup>`);
    expect(lookup.disabled).toBeTruthy();
    expect(lookup.readonly).toBeFalsy();
  });

  it('renders with readonly', () => {
    lookup = createFromTemplate(lookup, `<ids-lookup id="dropdown-1" readonly="true" label="Test"></ids-lookup>`);
    expect(lookup.readonly).toBeTruthy();
    expect(lookup.disabled).toBeFalsy();
  });

  it('should be able to set label', () => {
    expect(lookup.triggerField.label).toEqual('Normal Lookup');
    expect(lookup.label).toEqual('Normal Lookup');
    lookup.label = 'Test New Label';
    expect(lookup.triggerField.label).toEqual('Test New Label');
    expect(lookup.label).toEqual('Test New Label');
    lookup.label = '';
    expect(lookup.triggerField.label).toEqual('');
    expect(lookup.label).toEqual('');
  });

  it('should be able to set value', () => {
    lookup.value = '218901';
    expect(lookup.value).toEqual('218901');
    expect(lookup.triggerField.value).toEqual('218901');
  });

  it('should be able to set readonly with the property', () => {
    lookup.readonly = true;
    expect(lookup.triggerField.readonly).toEqual(true);
    expect(lookup.getAttribute('readonly')).toEqual('true');

    lookup.readonly = false;
    expect(lookup.triggerField.readonly).toEqual(false);
    expect(lookup.getAttribute('readonly')).toEqual(null);

    lookup.triggerField = null;
    expect(lookup.readonly).toEqual(false);
  });

  it('should skip readonly if no trigger field', () => {
    lookup.triggerField.remove();
    lookup.triggerField = null;
    lookup.readonly = true;
    expect(lookup.getAttribute('readonly')).toEqual(null);
    expect(lookup.readonly).toEqual(false);
  });

  it('should be able to set readonly with the attribute', () => {
    lookup.setAttribute('readonly', 'true');
    expect(lookup.triggerField.readonly).toEqual(true);
    expect(lookup.readonly).toEqual(true);

    lookup.setAttribute('readonly', 'false');
    expect(lookup.triggerField.readonly).toEqual(false);
    expect(lookup.readonly).toEqual(false);

    lookup.setAttribute('readonly', 'true');
    lookup.removeAttribute('readonly');
    expect(lookup.triggerField.readonly).toEqual(false);
    expect(lookup.readonly).toEqual(false);
  });

  it('should be able to set disabled with the property', () => {
    lookup.disabled = true;
    expect(lookup.triggerField.disabled).toEqual(true);
    expect(lookup.getAttribute('disabled')).toEqual('true');

    lookup.disabled = false;
    expect(lookup.triggerField.disabled).toEqual(false);
    expect(lookup.getAttribute('disabled')).toEqual(null);

    lookup.triggerField = null;
    expect(lookup.disabled).toEqual(false);
  });

  it('should be able to set disabled with the attribute', () => {
    lookup.setAttribute('disabled', 'true');
    expect(lookup.triggerField.disabled).toEqual(true);
    expect(lookup.disabled).toEqual(true);

    lookup.setAttribute('disabled', 'false');
    expect(lookup.triggerField.disabled).toEqual(false);
    expect(lookup.disabled).toEqual(false);

    lookup.setAttribute('disabled', 'true');
    lookup.removeAttribute('disabled');
    expect(lookup.triggerField.disabled).toEqual(false);
    expect(lookup.disabled).toEqual(false);
  });

  it('should be able to set tabbable with the property', () => {
    lookup.tabbable = true;
    expect(lookup.triggerField.tabbable).toEqual(true);
    expect(lookup.getAttribute('tabbable')).toEqual('true');

    lookup.tabbable = false;
    expect(lookup.triggerField.tabbable).toEqual(false);
    expect(lookup.getAttribute('tabbable')).toEqual('false');

    lookup.triggerField = null;
    lookup.tabbable = false;
    expect(lookup.tabbable).toEqual(false);
  });

  it('should be able to set tabbable with the attribute', () => {
    lookup.setAttribute('tabbable', 'true');
    expect(lookup.triggerField.tabbable).toEqual(true);
    expect(lookup.tabbable).toEqual(true);

    lookup.setAttribute('tabbable', 'false');
    expect(lookup.triggerField.tabbable).toEqual(false);
    expect(lookup.tabbable).toEqual(false);

    lookup.setAttribute('tabbable', 'true');
    lookup.removeAttribute('tabbable');
    expect(lookup.triggerField.tabbable).toEqual(false);
    expect(lookup.tabbable).toEqual(false);
  });

  it('should fire change on setting the value', () => {
    lookup.addEventListener('change', (e: CustomEvent) => {
      expect(e.detail.value).toEqual('218902');
    });
    lookup.value = '218902';
    expect(lookup.triggerField.value).toEqual('218902');
  });

  it('should be able to set clearable', () => {
    lookup.value = '102';
    expect(lookup.value).toEqual('102');
    expect(lookup.triggerField.value).toEqual('102');
    let clearableBtn = lookup.triggerField.container.querySelector('[part="clearable-button"]');
    expect(clearableBtn).toBeTruthy();
    clearableBtn.click();
    expect(lookup.value).toEqual('');
    expect(lookup.triggerField.value).toEqual('');

    lookup.triggerField.value = '102,10';
    expect(lookup.value).toEqual('102,10');
    expect(lookup.triggerField.value).toEqual('102,10');

    lookup.value = '102';
    lookup.clearable = false;
    clearableBtn = lookup.triggerField.container.querySelector('[part="clearable-button"]');
    expect(clearableBtn).not.toBeTruthy();

    lookup.clearable = true;
    clearableBtn = lookup.triggerField.container.querySelector('[part="clearable-button"]');
    expect(clearableBtn).toBeTruthy();
  });

  it('should open on click and close on the modal buttons', async () => {
    lookup = await createMultiSelectLookup();
    expect(lookup.modal.visible).toBe(false);
    lookup.triggerButton.click();
    expect(lookup.modal.visible).toBe(true);
    lookup.modal.buttons[0].click();
    expect(lookup.modal.visible).toBe(false);
    lookup.triggerButton.click();
    expect(lookup.modal.visible).toBe(true);
    lookup.modal.buttons[1].click();
    expect(lookup.modal.visible).toBe(false);
  });

  it('should open on down arrow', async () => {
    lookup = await createMultiSelectLookup();
    expect(lookup.modal.visible).toBe(false);
    const event = new KeyboardEvent('keydown', { key: 'ArrowDown' });
    lookup.dispatchEvent(event);
    expect(lookup.modal.visible).toBe(true);
    lookup.modal.buttons[0].click();
    expect(lookup.modal.visible).toBe(false);
  });

  it('should not open on click if readonly / disabled', () => {
    expect(lookup.modal.visible).toBe(false);
    lookup.readonly = true;
    lookup.triggerButton.click();
    expect(lookup.modal.visible).toBe(false);
    lookup.disabled = true;
    lookup.triggerButton.click();
    expect(lookup.modal.visible).toBe(false);
  });

  it('should be able to get and set readonly / disabled', () => {
    lookup.readonly = true;
    expect(lookup.readonly).toBe(true);
    expect(lookup.disabled).toBe(false);
    lookup.disabled = true;
    expect(lookup.readonly).toBe(false);
    expect(lookup.disabled).toBe(true);

    // Prevent error if no triggerField
    lookup.triggerField = null;
    lookup.disabled = false;
    expect(lookup.disabled).toBe(true);
  });

  it('should be able to set modal columns and dataset', () => {
    lookup.columns = columns();
    lookup.data = dataset;

    expect(lookup.dataGrid.shadowRoot.querySelectorAll('.ids-data-grid-row').length).toEqual(10);
    expect(lookup.dataGrid.columns.length).toEqual(6);
    expect(lookup.columns.length).toEqual(6);
    expect(lookup.data.length).toEqual(9);
    expect(lookup.dataGrid.shadowRoot.querySelectorAll('.ids-data-grid-cell').length).toEqual(54);
  });

  it('should be able to set dataGrid settings', () => {
    lookup.columns = columns();
    lookup.data = dataset;
    lookup.dataGridSettings = {
      rowHeight: 'sm'
    };

    expect(lookup.dataGrid.rowHeight).toEqual('sm');
    expect(lookup.dataGridSettings).toEqual({
      rowHeight: 'sm'
    });
  });

  it('renders with validation', () => {
    lookup = createFromTemplate(lookup, `<ids-lookup id="lookup-1" validate="required" validation-events="blur change" label="Test"></ids-lookup>`);
    expect(lookup.validate).toEqual('required');
    expect(lookup.validationEvents).toEqual('blur change');

    // Generate from the parent defaults
    lookup = createFromTemplate(lookup, `<ids-lookup id="lookup-1" validate="required" label="Test"></ids-lookup>`);
    lookup.validationEvents = 'blur change';
    expect(lookup.validate).toEqual('required');
    expect(lookup.validationEvents).toEqual('blur change');

    // Default Case
    lookup = createFromTemplate(lookup, `<ids-lookup id="lookup-1" validate="required" label="Test"></ids-lookup>`);
    expect(lookup.validate).toEqual('required');
    expect(lookup.validationEvents).toEqual('change blur');
  });

  it('supports changing validation dynamically', async () => {
    lookup = createFromTemplate(lookup, `<ids-lookup id="lookup-5" label="Dynamic Validation"></ids-lookup>`);
    await waitForTimeout(() => expect(lookup.shadowRoot.querySelector('ids-trigger-field')).toBeTruthy());

    (document.querySelector('ids-lookup') as any).validate = 'required';
    const triggerElem = lookup.shadowRoot.querySelector('ids-trigger-field');
    expect(triggerElem.getAttribute('validate')).toEqual('required');
    expect(triggerElem.getAttribute('validation-events')).toEqual('change blur');
    expect(triggerElem.labelEl).not.toEqual(undefined);

    (document.querySelector('ids-lookup') as any).validate = '';
    expect(triggerElem.getAttribute('validate')).toEqual(null);
    expect(triggerElem.getAttribute('validation-events')).toEqual(null);
    expect(triggerElem.labelEl).not.toEqual(undefined);

    expect(lookup.getAttribute('validate')).toEqual(null);
  });

  it('supports validation', async () => {
    lookup = createFromTemplate(lookup, `<ids-lookup id="lookup-5" label="Dropdown with Icons" validate="true">
     </ids-lookup>`);
    await waitForTimeout(() => expect(lookup.shadowRoot.querySelector('ids-trigger-field')).toBeTruthy());

    lookup.validate = 'required';
    lookup.validationEvents = 'blur change';
    lookup.triggerEvent('change', lookup);
    expect(lookup.getAttribute('validate')).toEqual('required');
  });

  it('can reset validation and validation-events', async () => {
    lookup.validate = 'required';
    lookup.validationEvents = 'blur change';
    lookup.validate = null;
    lookup.validationEvents = null;
    expect(lookup.getAttribute('validate')).toBeFalsy();
    expect(lookup.getAttribute('validation-events')).toBeFalsy();
  });

  it('should set selected rows on init', async () => {
    lookup = await createFromTemplate(lookup, `<ids-lookup id="lookup-1" label="Normal Lookup" title="Select an Item" field="description" value="102,103"></ids-lookup>`);
    lookup.columns = columns();
    lookup.data = dataset;
    expect(lookup.dataGrid.selectedRows.length).toEqual(2);
  });

  it('should default the field to id', () => {
    expect(lookup.field).toEqual('id');
  });

  it('should be able to set the delimiter', () => {
    expect(lookup.delimiter).toEqual(',');
    lookup.delimiter = '|';
    expect(lookup.delimiter).toEqual('|');
  });

  it('should be able to select two rows from the modal', async () => {
    lookup = await createMultiSelectLookup();
    expect(lookup.value).toEqual('');
    lookup.modal.visible = true;

    lookup.dataGrid.shadowRoot.querySelector('.ids-data-grid-body .ids-data-grid-row:nth-child(1) .ids-data-grid-cell:nth-child(1)').click();
    lookup.dataGrid.shadowRoot.querySelector('.ids-data-grid-body .ids-data-grid-row:nth-child(8) .ids-data-grid-cell:nth-child(1)').click();
    lookup.modal.buttons[1].click();

    expect(lookup.dataGrid.selectedRows.length).toEqual(2);
    expect(lookup.value).toEqual('101,108');
  });

  it('should fire selectionchanged event', async () => {
    const mockCallback = jest.fn();

    lookup = await createMultiSelectLookup();
    lookup.addEventListener('selectionchanged', mockCallback);

    lookup.modal.visible = true;
    lookup.dataGrid.shadowRoot.querySelector('.ids-data-grid-body .ids-data-grid-row:nth-child(8) .ids-data-grid-cell:nth-child(1) span').click();

    expect(mockCallback.mock.calls.length).toBe(1);
  });

  it('should fire rowdeselected event', async () => {
    const mockCallback = jest.fn((x) => {
      expect(x.detail.data.description).toEqual('108');
    });

    lookup = await createMultiSelectLookup();
    lookup.addEventListener('rowdeselected', mockCallback);

    lookup.modal.visible = true;
    lookup.dataGrid.shadowRoot.querySelector('.ids-data-grid-body .ids-data-grid-row:nth-child(8) .ids-data-grid-cell:nth-child(1)').click();
    lookup.dataGrid.shadowRoot.querySelector('.ids-data-grid-body .ids-data-grid-row:nth-child(8) .ids-data-grid-cell:nth-child(1)').click();

    expect(mockCallback.mock.calls.length).toBe(1);
  });

  it('supports custom modals', async () => {
    lookup = await createFromTemplate(lookup, `<ids-lookup id="lookup-5" label="Custom Lookup">
      <ids-modal slot="lookup-modal" id="custom-lookup-modal" aria-labelledby="custom-lookup-modal-title">
        <ids-text slot="title" font-size="24" type="h2" id="lookup-modal-title">Custom Lookup Modal</ids-text>
        <ids-modal-button slot="buttons" id="modal-cancel-btn" appearance="primary">
          <span>Apply</span>
        </ids-modal-button>
      </ids-modal>
    </ids-lookup>`);
    lookup.dataGridSettings = {
      rowSelection: 'multiple'
    };
    lookup.columns = columns();
    lookup.data = dataset;

    expect((document.querySelector('#custom-lookup-modal') as any).visible).toBeFalsy();
    lookup.modal.visible = true;
    expect((document.querySelector('#custom-lookup-modal') as any).visible).toBeTruthy();
  });

  it('supports setting mode', () => {
    lookup.mode = 'dark';
    expect(lookup.container.getAttribute('mode')).toEqual('dark');
  });

  it('supports setting autocomplete', () => {
    lookup.autocomplete = true;
    expect(lookup.hasAttribute('autocomplete')).toBeTruthy();
    lookup.autocomplete = false;
    expect(lookup.hasAttribute('autocomplete')).toBeFalsy();
  });

  it('supports setting dirty tracker', () => {
    expect(lookup.triggerField.shadowRoot.querySelector('.icon-dirty')).toBeFalsy();
    lookup.value = '';
    lookup.dirtyTracker = true;
    lookup.value = '100';
    expect(lookup.triggerField.shadowRoot.querySelector('.icon-dirty')).toBeTruthy();
    lookup.value = '';
    lookup.onDirtyTrackerChange(false);
    expect(lookup.triggerField.shadowRoot.querySelector('.icon-dirty')).toBeFalsy();
  });

  it('should be able to set attributes before append', async () => {
    const elem: any = new IdsLookup();
    elem.id = 'test';
    elem.value = '102,103';
    elem.dirtyTracker = true;
    elem.autocomplete = true;
    document.body.appendChild(elem);

    expect(elem.getAttribute('id')).toEqual('test');
    expect(elem.getAttribute('value')).toEqual('102,103');
    expect(elem.getAttribute('dirty-tracker')).toEqual('true');
  });

  it('should be able to set attributes after append', async () => {
    const elem: any = new IdsLookup();
    document.body.appendChild(elem);
    elem.id = 'test';
    elem.value = '102,103';
    elem.dirtyTracker = true;

    expect(elem.getAttribute('id')).toEqual('test');
    expect(elem.getAttribute('value')).toEqual('102,103');
    expect(elem.getAttribute('dirty-tracker')).toEqual('true');
  });

  it('should render field height', () => {
    const heights = ['xs', 'sm', 'md', 'lg'];
    const defaultHeight = 'md';
    const className = (h: any) => `field-height-${h}`;
    const checkHeight = (height: any) => {
      lookup.fieldHeight = height;

      expect(lookup.input.getAttribute('field-height')).toEqual(height);
      expect(lookup.container.classList).toContain(className(height));
      heights.filter((h) => h !== height).forEach((h) => {
        expect(lookup.container.classList).not.toContain(className(h));
      });
    };

    expect(lookup.getAttribute('field-height')).toEqual(null);
    heights.filter((h) => h !== defaultHeight).forEach((h) => {
      expect(lookup.container.classList).not.toContain(className(h));
    });

    expect(lookup.container.classList).toContain(className(defaultHeight));
    heights.forEach((h) => checkHeight(h));
    lookup.removeAttribute('field-height');
    lookup.removeAttribute('compact');

    expect(lookup.getAttribute('field-height')).toEqual(null);
    heights.filter((h) => h !== defaultHeight).forEach((h) => {
      expect(lookup.container.classList).not.toContain(className(h));
    });
    lookup.onFieldHeightChange();

    expect(lookup.container.classList).toContain(className(defaultHeight));
  });

  it('should set compact height', () => {
    lookup.compact = true;

    expect(lookup.hasAttribute('compact')).toBeTruthy();
    expect(lookup.container.classList.contains('compact')).toBeTruthy();
    lookup.compact = false;

    expect(lookup.hasAttribute('compact')).toBeFalsy();
    expect(lookup.container.classList.contains('compact')).toBeFalsy();
  });

  it('should set size', () => {
    const sizes = ['xs', 'sm', 'mm', 'md', 'lg', 'full'];
    const defaultSize = 'md';
    const checkSize = (size: any) => {
      lookup.size = size;

      expect(lookup.getAttribute('size')).toEqual(size);
      expect(lookup.input.getAttribute('size')).toEqual(size);
    };

    expect(lookup.getAttribute('size')).toEqual(null);
    expect(lookup.input.getAttribute('size')).toEqual(defaultSize);
    sizes.forEach((s) => checkSize(s));
    lookup.size = null;

    expect(lookup.getAttribute('size')).toEqual(null);
    expect(lookup.input.getAttribute('size')).toEqual(defaultSize);
  });
});
