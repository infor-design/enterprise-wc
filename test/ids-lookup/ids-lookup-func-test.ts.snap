// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsLookup Component renders correctly 1`] = `"<ids-lookup id="lookup-1" label="Normal Lookup"></ids-lookup>"`;

exports[`IdsLookup Component renders correctly 2`] = `
"<style nonce="0a59a005">:host { background-color: transparent; }</style>
    <ids-trigger-field label="Normal Lookup" part="trigger-field" value="" clearable="true" size="md" field-height="md" class="field-height-md">
      <ids-trigger-button slot="trigger-end" part="trigger-lookup" tabbable="" field-height="md">
        <ids-text audible="true">LookupTriggerButton</ids-text>
        <ids-icon icon="search-list" part="icon"></ids-icon>
      </ids-trigger-button>
    </ids-trigger-field>
    <slot name="lookup-modal">
      <ids-modal id="lookup-modal" aria-labelledby="lookup-modal-title" part="modal" message-title="Normal Lookup" aria-label="Normal Lookup" role="dialog" fullsize="sm" respond-down="sm">
        <ids-text slot="title" font-size="24" type="h2" id="lookup-modal-title">Normal Lookup</ids-text>
        <ids-layout-grid class="data-grid-container" auto="true" gap="md" no-margins="true" min-col-width="600px">
          <ids-layout-grid-cell>
            <ids-data-grid id="lookup-data-grid" label="Normal Lookup" part="data-grid" list-style="true">
            </ids-data-grid>
          </ids-layout-grid-cell>
        </ids-layout-grid>

      <ids-modal-button slot="buttons" id="modal-cancel-btn" appearance="secondary">
        <span>Cancel</span>
        </ids-modal-button>
        <ids-modal-button slot="buttons" id="modal-apply-btn" appearance="primary">
          <span>Apply</span>
        </ids-modal-button>
      </ids-modal>
    </slot>
    "
`;
